File: 241002/승자독식 모노폴리/odd-monopoly.py
def int_minus_one():
    return lambda x: int(x) - 1


def initialize():
    global exclusive_contracts, player_pos
    for i in range(n):
        for j in range(n):
            if grid[i][j]:
                exclusive_contracts[i][j] = (k, grid[i][j])
                player_pos[grid[i][j]] = (i, j)


def players_move():
    temp = [[[] for _ in range(n)] for _ in range(n)]
    for pid in range(1, m + 1):
        if not alive[pid]:
            continue
        x, y = player_pos[pid]
        cur_d = player_d[pid - 1]
        player_d_priority = player_d_priorities[4 * (pid - 1) + cur_d]
        moved = False
        for d in player_d_priority:
            dx, dy = dirs[d]
            nx = x + dx
            ny = y + dy
            if 0 <= nx < n and 0 <= ny < n and exclusive_contracts[nx][ny] == (0, 0):
                temp[nx][ny].append(pid)
                player_pos[pid] = (nx, ny)
                player_d[pid - 1] = d
                moved = True
                break
        if moved:
            continue
        for d in player_d_priority:
            dx, dy = dirs[d]
            nx = x + dx
            ny = y + dy
            if 0 <= nx < n and 0 <= ny < n and exclusive_contracts[nx][ny][1] == pid:
                temp[nx][ny].append(pid)
                player_pos[pid] = (nx, ny)
                player_d[pid - 1] = d
                moved = True
                break
        if moved:
            continue
        temp[x][y].append(pid)
    for i in range(n):
        for j in range(n):
            if not temp[i][j]:
                continue
            if len(temp[i][j]) == 1:
                exclusive_contracts[i][j] = (k, temp[i][j][0])
                grid[i][j] = temp[i][j][0]
                continue
            min_player = min(temp[i][j])
            exclusive_contracts[i][j] = (k, min_player)
            grid[i][j] = min_player
            for temp_id in temp[i][j]:
                if temp_id != min_player:
                    delete_player(temp_id)


def delete_player(pid):
    global player_pos, alive
    del player_pos[pid]
    alive[pid] = 0


def decrease_time():
    global exclusive_contracts
    for i in range(n):
        for j in range(n):
            if exclusive_contracts[i][j] == (0, 0):
                continue
            time, pid = exclusive_contracts[i][j]
            time -= 1
            if time == 0:
                exclusive_contracts[i][j] = (0, 0)
            else:
                exclusive_contracts[i][j] = (time, pid)


def check_only_1st_player_alive():
    return sum(alive) == 1 and alive[1]


dirs = ((-1, 0), (1, 0), (0, -1), (0, 1))
n, m, k = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(n)]
exclusive_contracts = [[(0, 0) for _ in range(n)] for _ in range(n)]  # (time, pid)
player_pos = dict()
player_d = list(map(int_minus_one(), input().split()))
player_d_priorities = [tuple(map(int_minus_one(), input().split())) for _ in range(m * 4)]
alive = [0] + [1] * m
initialize()
turn_cnt = 1
for _ in range(1_000):
    players_move()
    if check_only_1st_player_alive():
        break
    decrease_time()
    turn_cnt += 1
print(turn_cnt if turn_cnt < 1_000 else -1)

File: 241003/승자독식 모노폴리/odd-monopoly.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader br;
	static StringTokenizer st;

	static final int[][] dirs = { { -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 } };
	static int n, m, k;
	static int[][] grid;
	static Contract[][] exclusiveContracts;
	static int[] playerDirs;
	static int[][] playerDirPriorities;
	static boolean[] gameOver;

	public static void main(String[] args) throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		grid = new int[n][n];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < n; j++) {
				grid[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		exclusiveContracts = new Contract[n][n];
		playerDirs = new int[m + 1];
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < m; i++) {
			playerDirs[i] = Integer.parseInt(st.nextToken()) - 1;
		}
		playerDirPriorities = new int[m * 4][4];
		for (int i = 0; i < m * 4; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < 4; j++) {
				playerDirPriorities[i][j] = Integer.parseInt(st.nextToken()) - 1;
			}
		}
		gameOver = new boolean[m + 1];
		int turnCount = 1;
		for (int i = 0; i < 1_000; i++) {
			playersContract();
			playersMove();
			decreaseTimes();
			if (checkOnlyOnePlayerAlive()) {
				break;
			}
			turnCount++;
		}
		if (turnCount >= 1_000) {
			turnCount = -1;
		}
		System.out.println(turnCount);
	}

	private static void playersContract() {
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				if (grid[i][j] > 0) {
					exclusiveContracts[i][j] = new Contract(k, grid[i][j]);
				}
			}
		}
	}

	private static void playersMove() {
		final List<Integer>[][] temp = new List[n][n];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				temp[i][j] = new ArrayList<>();
			}
		}
		for (int x = 0; x < n; x++) {
			for (int y = 0; y < n; y++) {
				if (grid[x][y] == 0) {
					continue;
				}
				final int pid = grid[x][y];
				final int curDir = playerDirs[pid - 1];
				final int[] playerDirPriority = playerDirPriorities[4 * (pid - 1) + curDir];
				boolean moved = false;
				for (final int d : playerDirPriority) {
					final int nx = x + dirs[d][0];
					final int ny = y + dirs[d][1];
					if (0 <= nx && nx < n && 0 <= ny && ny < n && exclusiveContracts[nx][ny] == null) {
						temp[nx][ny].add(pid);
						playerDirs[pid - 1] = d;
						moved = true;
						break;
					}
				}
				if (moved) {
					continue;
				}
				for (final int d : playerDirPriority) {
					final int nx = x + dirs[d][0];
					final int ny = y + dirs[d][1];
					if (0 <= nx && nx < n && 0 <= ny && ny < n && exclusiveContracts[nx][ny].isSamePID(pid)) {
						temp[nx][ny].add(pid);
						playerDirs[pid - 1] = d;
						break;
					}
				}
			}
		}
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				if (temp[i][j].isEmpty()) {
					grid[i][j] = 0;
					continue;
				}
				if (temp[i][j].size() == 1) {
					grid[i][j] = temp[i][j].get(0);
					continue;
				}
				final int minPlayer = Collections.min(temp[i][j]);
				grid[i][j] = minPlayer;
				for (final int tempId : temp[i][j]) {
					if (tempId != minPlayer) {
						gameOver[tempId] = true;
					}
				}
			}
		}
	}

	private static void decreaseTimes() {
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				if (exclusiveContracts[i][j] == null) {
					continue;
				}
				final Contract contract = exclusiveContracts[i][j];
				contract.decreaseTime();
				if (contract.isTimeZero()) {
					exclusiveContracts[i][j] = null;
				}
			}
		}
	}

	private static boolean checkOnlyOnePlayerAlive() {
		int count = 0;
		for (int i = 1; i <= m; i++) {
			if (!gameOver[i]) {
				count++;
			}
		}
		return count == 1;
	}

	static class Contract {

		int time;
		int pid;

		public Contract(int time, int pid) {
			super();
			this.time = time;
			this.pid = pid;
		}

		public void decreaseTime() {
			time--;
		}

		public boolean isTimeZero() {
			return time == 0;
		}

		public boolean isSamePID(final int pid) {
			return this.pid == pid;
		}
	}
}